File: kernel/timer.c
// SPDX-License-Identifier: GPL-3.0-or-later
/*
 * SlipperKernel — S-mode timer driven by the CLINT.
 *
 * We program mtimecmp directly via MMIO (no SBI calls). The CLINT lives
 * at 0x02000000 on QEMU virt.
 *
 * Tick rate: 100 Hz.
 */
#include "types.h"
#include "timer.h"
#include "klog.h"
#include "riscv.h"
#include "fdt.h"
#include "proc.h"

#define CLINT_TIMER_HZ   100

static volatile u64 *g_mtime    = NULL;
static volatile u64 *g_mtimecmp = NULL;
static u64  g_tick_interval     = 0;
static u64  g_uptime_ticks      = 0;
volatile u64 g_jiffies          = 0;
static paddr_t g_clint_base     = 0;

static inline u64 read_mtime(void)
{
    /* Guard against non-atomic 64-bit MMIO read by retrying on low/high
     * mismatch. On QEMU this is a single 64-bit load and always returns
     * cleanly, but real boards (SG2000) may return torn reads. */
    u32 hi, lo, hi2;
    do {
        hi  = (u32)(*g_mtime >> 32);
        lo  = (u32)(*g_mtime & 0xFFFFFFFF);
        hi2 = (u32)(*g_mtime >> 32);
    } while (hi != hi2);
    return ((u64)hi << 32) | lo;
}

static inline void write_mtimecmp(u64 v)
{
    /* Per spec: write high word to 0xFFFFFFFF, then low word, then high
     * word, to avoid spurious interrupts from a transient low value. */
    volatile u32 *p = (volatile u32 *)g_mtimecmp;
    p[1] = 0xFFFFFFFF;
    p[0] = (u32)v;
    p[1] = (u32)(v >> 32);
}

void timer_init(void)
{
    fdt_find_clint(&g_clint_base);
    g_mtime    = (volatile u64 *)(g_clint_base + 0xBFF8);
    g_mtimecmp = (volatile u64 *)(g_clint_base + 0x4000);   /* hart 0 */

    /* QEMU virt CLINT runs at 10 MHz. Compute interval for CLINT_TIMER_HZ. */
    u64 freq = 10 * 1000 * 1000;
    g_tick_interval = freq / CLINT_TIMER_HZ;

    /* First deadline. */
    write_mtimecmp(read_mtime() + g_tick_interval);

    /* Enable S-mode timer interrupt:
     *   sie.STIE  (S-mode interrupt-enable for S-timer)
     *   sip.STIP  is read-only set by CLINT.
     * We are in S-mode, so we use sie (not mie which is M-mode only).
     * Globally enabling sstatus.SIE is done in kmain before proc_enter_user. */
    csr_set(sie, 1U << 5);

    kinf("timer: CLINT @0x%lx, tick=%lu ns",
         g_clint_base, g_tick_interval * 1000ULL / (freq / 1000000));
}

void timer_handle(void)
{
    g_uptime_ticks++;
    g_jiffies++;
    /* Re-arm. */
    write_mtimecmp(read_mtime() + g_tick_interval);
    sched_tick();
}

u64 timer_now(void)
{
    return read_mtime();
}

u64 timer_uptime_us(void)
{
    /* CLINT @10MHz → 1 tick = 100 ns = 0.1 µs */
    return g_uptime_ticks * (1000 * 1000 / CLINT_TIMER_HZ);
}

File: kernel/src/kernel/timer.rs
//! CLINT timer (100 Hz tick).
use crate::arch::regs::*;
use crate::arch::{csr, mmio::Mmio};
use crate::proc::proc;
static mut G_MTIME: usize = 0;
static mut G_MTIMECMP: usize = 0;
static mut G_FREQ: u64 = CLINT_FREQ_QEMU;
static mut G_TICK_INTERVAL: u64 = 0;
static mut G_UPTICKS: u64 = 0;
pub static mut G_JIFFIES: u64 = 0;

pub unsafe fn init() {
    let clint = CLINT_BASE;
    G_MTIME = (clint + 0xBFF8) as usize;
    G_MTIMECMP = (clint + 0x4000) as usize;
    G_FREQ = CLINT_FREQ_QEMU;
    G_TICK_INTERVAL = G_FREQ / 100;
    let now = read_mtime();
    write_mtimecmp(now + G_TICK_INTERVAL);
    csr::set_sie(1 << 5);
    crate::kinf!(
        "timer",
        "CLINT @%p, tick=%d ns",
        onyx_core::fmt::Arg::from(clint),
        onyx_core::fmt::Arg::from(1_000_000_000u64 / G_FREQ)
    );
}

unsafe fn read_mtime() -> u64 {
    loop {
        let hi = Mmio::<u32>::at(G_MTIME + 4).read();
        let lo = Mmio::<u32>::at(G_MTIME).read();
        let hi2 = Mmio::<u32>::at(G_MTIME + 4).read();
        if hi == hi2 {
            return ((hi as u64) << 32) | (lo as u64);
        }
    }
}

unsafe fn write_mtimecmp(v: u64) {
    Mmio::<u32>::at(G_MTIMECMP + 4).write(0xFFFF_FFFF);
    Mmio::<u32>::at(G_MTIMECMP).write(v as u32);
    Mmio::<u32>::at(G_MTIMECMP + 4).write((v >> 32) as u32);
}

pub unsafe fn handle() {
    G_UPTICKS = G_UPTICKS.wrapping_add(1);
    G_JIFFIES = G_JIFFIES.wrapping_add(1);
    let now = read_mtime();
    write_mtimecmp(now + G_TICK_INTERVAL);
    proc::sched_tick();
}
pub fn uptime_us() -> u64 {
    unsafe { G_UPTICKS * 10_000 }
}
